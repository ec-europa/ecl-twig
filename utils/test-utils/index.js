const Twig = require('twig');
const { format } = require('prettier');

const renderTwigFile = (file, options, cb) =>
  Twig.renderFile(file, options, (err, html) =>
    cb(err, format(html, { parser: 'html' }))
  );

const renderTwigFileAsNode = (file, options) =>
  new Promise((resolve, reject) =>
    Twig.renderFile(file, options, (err, html) => {
      if (err) return reject(err);

      const div = document.createElement('div');
      div.innerHTML = html.trim();

      return resolve(div.firstChild);
    })
  );

module.exports = {
  renderTwigFile,
  renderTwigFileAsNode,
};
