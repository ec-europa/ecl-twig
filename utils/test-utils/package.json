{
  "name": "@ecl-twig/test-utils",
  "version": "2.29.0",
  "description": "A local test helper",
  "main": "index.js",
  "private": true,
  "devDependencies": {
    "deepmerge": "4.2.2"
  }
}
