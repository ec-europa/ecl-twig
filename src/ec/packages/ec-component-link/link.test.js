import { merge, renderTwigFileAsNode } from '@ecl-twig/test-utils';

describe('EC - Link', () => {
  const template = '@ecl-twig/ec-component-link/ecl-link.html.twig';
  const render = params => renderTwigFileAsNode(template, params);
  const defaultIconPath = 'static/icons.svg';
  const defaultDataStructure = {
    link: {
      type: '',
      label: '',
      path: '/path',
    },
  };

  describe('Default', () => {
    test('renders correctly', () => {
      expect.assertions(1);

      const options = merge(defaultDataStructure, {
        link: {
          type: 'default',
          label: 'Default link',
        },
      });

      return expect(render(options)).resolves.toMatchSnapshot();
    });
  });

  describe('Standalone', () => {
    test('renders correctly', () => {
      expect.assertions(1);

      const options = merge(defaultDataStructure, {
        link: {
          type: 'standalone',
          label: 'Standalone link',
        },
      });

      return expect(render(options)).resolves.toMatchSnapshot();
    });
  });

  describe('Standalone with missing icon name', () => {
    const options = merge(defaultDataStructure, {
      link: {
        type: 'standalone',
        label: 'Standalone link with icon',
        icon_position: 'after',
      },
      icon: {
        type: 'ui',
        size: 'fluid',
        path: defaultIconPath,
      },
    });

    test('renders correctly', () => {
      expect.assertions(1);
      return expect(render(options)).resolves.toMatchSnapshot();
    });
  });

  describe('With two icons after', () => {
    const options = merge(defaultDataStructure, {
      link: {
        type: 'standalone',
        label: 'Standalone link with icon',
        icon_position: 'after',
      },
      icon: [
        {
          type: 'ui',
          name: 'external',
          size: 'fluid',
          path: defaultIconPath,
        },
        {
          type: 'ui',
          name: 'download',
          size: 'fluid',
          path: defaultIconPath,
        },
      ],
    });

    test('renders correctly', () => {
      expect.assertions(1);
      return expect(render(options)).resolves.toMatchSnapshot();
    });

    test('renders correctly with extra class names', () => {
      expect.assertions(1);

      const withExtraClasses = merge(options, {
        extra_classes: 'custom-class custom-class--test',
      });

      return expect(render(withExtraClasses)).resolves.toMatchSnapshot();
    });
  });

  describe('With missing icon name', () => {
    const options = merge(defaultDataStructure, {
      link: {
        type: 'standalone',
        label: 'Standalone link with icon',
        icon_position: 'after',
      },
      icon: [
        {
          type: 'ui',
          name: 'external',
          size: 'fluid',
          path: defaultIconPath,
        },
        {
          type: 'ui',
          size: 'xs',
          path: defaultIconPath,
        },
      ],
    });

    test('renders correctly', () => {
      expect.assertions(1);
      return expect(render(options)).resolves.toMatchSnapshot();
    });
  });

  describe('With two icons before', () => {
    const options = merge(defaultDataStructure, {
      link: {
        type: 'standalone',
        label: 'Standalone link with icon',
        icon_position: 'before',
      },
      icon: [
        {
          type: 'ui',
          name: 'external',
          size: 'fluid',
          path: defaultIconPath,
        },
        {
          type: 'ui',
          name: 'download',
          size: 'fluid',
          path: defaultIconPath,
        },
      ],
    });

    test('renders correctly', () => {
      expect.assertions(1);
      return expect(render(options)).resolves.toMatchSnapshot();
    });

    test('renders correctly with extra class names', () => {
      expect.assertions(1);

      const withExtraClasses = merge(options, {
        extra_classes: 'custom-class custom-class--test',
      });

      return expect(render(withExtraClasses)).resolves.toMatchSnapshot();
    });
  });

  describe('With icon before', () => {
    test('renders correctly', () => {
      expect.assertions(1);

      const options = merge(defaultDataStructure, {
        link: {
          type: 'standalone',
          label: 'Standalone link with icon',
          icon_position: 'before',
        },
        icon: {
          type: 'ui',
          name: 'external',
          size: 'fluid',
          path: defaultIconPath,
          extra_classes: 'ecl-test-extra-class',
        },
      });

      return expect(render(options)).resolves.toMatchSnapshot();
    });
  });

  describe('With icon after', () => {
    const options = merge(defaultDataStructure, {
      link: {
        type: 'standalone',
        label: 'Standalone link with icon',
        icon_position: 'after',
      },
      icon: {
        type: 'ui',
        name: 'external',
        size: 'fluid',
        path: defaultIconPath,
        extra_classes: 'ecl-test-extra-class',
      },
    });

    test('renders correctly', () => {
      expect.assertions(1);
      return expect(render(options)).resolves.toMatchSnapshot();
    });

    test('renders correctly with extra class names', () => {
      expect.assertions(1);

      const withExtraClasses = merge(options, {
        extra_classes: 'custom-class custom-class--test',
      });

      return expect(render(withExtraClasses)).resolves.toMatchSnapshot();
    });

    test('renders correctly with extra attributes', () => {
      expect.assertions(1);

      const withExtraAttributes = merge(options, {
        extra_attributes: [
          { name: 'data-test', value: 'data-test-value' },
          { name: 'data-test-1', value: 'data-test-value-1' },
        ],
      });

      return expect(render(withExtraAttributes)).resolves.toMatchSnapshot();
    });
  });

  describe('Cta variant with icon', () => {
    const options = merge(defaultDataStructure, {
      link: {
        type: 'cta',
        label: 'Call to action link',
        icon_position: 'after',
      },
      icon: {
        type: 'ui',
        name: 'corner-arrow',
        size: 'fluid',
        path: defaultIconPath,
        transform: 'rotate-90',
      },
    });

    test('renders correctly', () => {
      expect.assertions(1);
      return expect(render(options)).resolves.toMatchSnapshot();
    });
  });
});
