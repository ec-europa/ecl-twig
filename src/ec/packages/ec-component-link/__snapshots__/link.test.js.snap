// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`EC - Link Default renders correctly 1`] = `
<a
  class="ecl-link ecl-link--default"
  href="/path"
>
  Default link
</a>
`;

exports[`EC - Link Standalone renders correctly 1`] = `
<a
  class="ecl-link ecl-link--standalone"
  href="/path"
>
  Standalone link
</a>
`;

exports[`EC - Link Standalone with missing icon name renders correctly 1`] = `
<a
  class="ecl-link ecl-link--standalone"
  href="/path"
>
  Standalone link with icon
</a>
`;

exports[`EC - Link With icon after renders correctly 1`] = `
<a
  class="ecl-link ecl-link--standalone ecl-link--icon ecl-link--icon-after"
  href="/path"
>
  <span
    class="ecl-link__label"
  >
    Standalone link with icon
  </span>
   
  <svg
    aria-hidden="true"
    class="ecl-icon ecl-icon--fluid ecl-link__icon ecl-test-extra-class"
    focusable="false"
  >
    <use
      xlink:href="static/icons.svg#ui--external"
    />
  </svg>
</a>
`;

exports[`EC - Link With icon after renders correctly with extra attributes 1`] = `
<a
  class="ecl-link ecl-link--standalone ecl-link--icon ecl-link--icon-after"
  data-test="data-test-value"
  data-test-1="data-test-value-1"
  href="/path"
>
  <span
    class="ecl-link__label"
  >
    Standalone link with icon
  </span>
   
  <svg
    aria-hidden="true"
    class="ecl-icon ecl-icon--fluid ecl-link__icon ecl-test-extra-class"
    focusable="false"
  >
    <use
      xlink:href="static/icons.svg#ui--external"
    />
  </svg>
</a>
`;

exports[`EC - Link With icon after renders correctly with extra class names 1`] = `
<a
  class="ecl-link ecl-link--standalone ecl-link--icon ecl-link--icon-after custom-link custom-link--test"
  href="/path"
>
  <span
    class="ecl-link__label"
  >
    Standalone link with icon
  </span>
   
  <svg
    aria-hidden="true"
    class="ecl-icon ecl-icon--fluid ecl-link__icon ecl-test-extra-class"
    focusable="false"
  >
    <use
      xlink:href="static/icons.svg#ui--external"
    />
  </svg>
</a>
`;

exports[`EC - Link With icon before renders correctly 1`] = `
<a
  class="ecl-link ecl-link--standalone ecl-link--icon ecl-link--icon-before"
  href="/path"
>
  <svg
    aria-hidden="true"
    class="ecl-icon ecl-icon--fluid ecl-link__icon ecl-test-extra-class"
    focusable="false"
  >
    <use
      xlink:href="static/icons.svg#ui--external"
    />
  </svg>
   
  <span
    class="ecl-link__label"
  >
    Standalone link with icon
  </span>
</a>
`;

exports[`EC - Link With missing icon name renders correctly 1`] = `
<a
  class="ecl-link ecl-link--standalone ecl-link--icon ecl-link--icon-after"
  href="/path"
>
  <span
    class="ecl-link__label"
  >
    Standalone link with icon
  </span>
   
  <svg
    aria-hidden="true"
    class="ecl-icon ecl-icon--fluid ecl-link__icon"
    focusable="false"
  >
    <use
      xlink:href="static/icons.svg#ui--external"
    />
  </svg>
</a>
`;

exports[`EC - Link With two icons after renders correctly 1`] = `
<a
  class="ecl-link ecl-link--standalone ecl-link--icon ecl-link--icon-after"
  href="/path"
>
  <span
    class="ecl-link__label"
  >
    Standalone link with icon
  </span>
   
  <svg
    aria-hidden="true"
    class="ecl-icon ecl-icon--fluid ecl-link__icon"
    focusable="false"
  >
    <use
      xlink:href="static/icons.svg#ui--external"
    />
  </svg>
  <svg
    aria-hidden="true"
    class="ecl-icon ecl-icon--fluid ecl-link__icon"
    focusable="false"
  >
    <use
      xlink:href="static/icons.svg#ui--download"
    />
  </svg>
</a>
`;

exports[`EC - Link With two icons after renders correctly with extra class names 1`] = `
<a
  class="ecl-link ecl-link--standalone ecl-link--icon ecl-link--icon-after custom-link custom-link--test"
  href="/path"
>
  <span
    class="ecl-link__label"
  >
    Standalone link with icon
  </span>
   
  <svg
    aria-hidden="true"
    class="ecl-icon ecl-icon--fluid ecl-link__icon"
    focusable="false"
  >
    <use
      xlink:href="static/icons.svg#ui--external"
    />
  </svg>
  <svg
    aria-hidden="true"
    class="ecl-icon ecl-icon--fluid ecl-link__icon"
    focusable="false"
  >
    <use
      xlink:href="static/icons.svg#ui--download"
    />
  </svg>
</a>
`;

exports[`EC - Link With two icons before renders correctly 1`] = `
<a
  class="ecl-link ecl-link--standalone ecl-link--icon ecl-link--icon-before"
  href="/path"
>
  <svg
    aria-hidden="true"
    class="ecl-icon ecl-icon--fluid ecl-link__icon"
    focusable="false"
  >
    <use
      xlink:href="static/icons.svg#ui--external"
    />
  </svg>
  <svg
    aria-hidden="true"
    class="ecl-icon ecl-icon--fluid ecl-link__icon"
    focusable="false"
  >
    <use
      xlink:href="static/icons.svg#ui--download"
    />
  </svg>
   
  <span
    class="ecl-link__label"
  >
    Standalone link with icon
  </span>
</a>
`;

exports[`EC - Link With two icons before renders correctly with extra class names 1`] = `
<a
  class="ecl-link ecl-link--standalone ecl-link--icon ecl-link--icon-before custom-link custom-link--test"
  href="/path"
>
  <svg
    aria-hidden="true"
    class="ecl-icon ecl-icon--fluid ecl-link__icon"
    focusable="false"
  >
    <use
      xlink:href="static/icons.svg#ui--external"
    />
  </svg>
  <svg
    aria-hidden="true"
    class="ecl-icon ecl-icon--fluid ecl-link__icon"
    focusable="false"
  >
    <use
      xlink:href="static/icons.svg#ui--download"
    />
  </svg>
   
  <span
    class="ecl-link__label"
  >
    Standalone link with icon
  </span>
</a>
`;
